src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php

```diff
@@ -42,51 +42,72 @@ public function __construct(Connection $connection, SerializerInterface $seriali
     public function receive(callable $handler): void
     {
         while (!$this->shouldStop) {
-            $AMQPEnvelope = $this->connection->get();
+            try {
+                $AMQPEnvelope = $this->connection->get();
+            } catch (\AMQPException $exception) {
+                throw new TransportException($exception->getMessage(), 0, $exception);
+            }
+
             if (null === $AMQPEnvelope) {
-                $handler(null);
+                $handler(null, null);
 
-                usleep($this->connection->getConnectionCredentials()['loop_sleep'] ?? 200000);
+                usleep($this->connection->getConnectionConfiguration()['loop_sleep'] ?? 200000);
                 if (\function_exists('pcntl_signal_dispatch')) {
                     pcntl_signal_dispatch();
                 }
 
                 continue;
             }
 
-            try {
-                $handler($this->serializer->decode([
-                    'body' => $AMQPEnvelope->getBody(),
-                    'headers' => $AMQPEnvelope->getHeaders(),
-                ]));
-
-                $this->connection->ack($AMQPEnvelope);
-            } catch (RejectMessageExceptionInterface $e) {
-                try {
-                    $this->connection->reject($AMQPEnvelope);
-                } catch (\AMQPException $exception) {
-                    throw new TransportException($exception->getMessage(), 0, $exception);
-                }
-
-                throw $e;
-            } catch (\AMQPException $e) {
-                throw new TransportException($e->getMessage(), 0, $e);
-            } catch (\Throwable $e) {
-                try {
-                    $this->connection->nack($AMQPEnvelope, AMQP_REQUEUE);
-                } catch (\AMQPException $exception) {
-                    throw new TransportException($exception->getMessage(), 0, $exception);
-                }
+            $handler($this->serializer->decode([
+                'body' => $AMQPEnvelope->getBody(),
+                'headers' => $AMQPEnvelope->getHeaders(),
+            ]), new QueuedMessageMetadata(
+                $AMQPEnvelope,
+                $AMQPEnvelope->getMessageId(),
+                (int) $AMQPEnvelope->getHeader(Connection::ATTEMPT_COUNT_HEADER_NAME) ?: 0
+            ));
 
-                throw $e;
-            } finally {
-                if (\function_exists('pcntl_signal_dispatch')) {
-                    pcntl_signal_dispatch();
-                }
+            if (\function_exists('pcntl_signal_dispatch')) {
+                pcntl_signal_dispatch();
             }
         }
     }
 
+    public function acknowledge($message): void
```

Comment:
IMHO these methods should take the envelope, so it's consistent with the other ones and could even be moved to an interface later.

Reply:
I've removed this arg internally, and they now rely on a "currentAmqpMessage". It's temporary state... but I think it's safe, and it keeps that internal `$message` "thing" (which will vary based on the transport) inside the transport.